docs/docs/ProgrammingGuide/add-layer-criterion.md

```diff
@@ -0,0 +1,236 @@
+type is Table, the gradInput buffer is also inited as an empty Table.
+
+## Create a layer with multiple outputs
+When a layer has multiple outputs, we just need to specify the output type as **Table**.
+
+Say we want a layer to split a N-d tensor into many (N - 1)-d tensors in the first dimension. Here
+is the code example:
+
+```scala
+import com.intel.analytics.bigdl.nn.abstractnn.AbstractModule
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.utils.{T, Table}
+
+class UnpackLayer extends AbstractModule[Tensor[Float], Table, Float]{
+
+  override def updateOutput(input: Tensor[Float]): Table = {
+    require(input.nDimension() > 1)
+    output = T()
+    (1 to input.size(1)).foreach(i => {
+      output(i) = Tensor[Float]().resizeAs(input.select(1, i)).copy(input.select(1, i))
+    })
+    output
+  }
+
+  override def updateGradInput(input: Tensor[Float], gradOutput: Table): Tensor[Float] = {
+    gradInput.resizeAs(input)
+    (1 to input.size(1)).foreach(i => {
+      gradInput.select(1, i).copy(gradOutput[Tensor[Float]](i))
+    })
+    gradInput
+  }
+}
+```
+
+
+## Create a layer with parameters
+Some layers are trainable, which means they have some parameters and the parameters are changed in
+the training process. Trainable layer need to override extra methods.
+
+Parameters are all tensors and their numeric type is determined by the third generic type of the
+AbstractModule.
+
+Say we want to create a tensor which multiply a N x 3 matrix with a 3 x 2 parameter, and get a N x 2
+output. It looks like our first example. Here is the initial code:
+```scala
+import com.intel.analytics.bigdl.nn.abstractnn.AbstractModule
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.utils.{T, Table}
+
+class MatrixLayer extends AbstractModule[Tensor[Float], Tensor[Float], Float]{
+
+  private val w = Tensor[Float](3, 2)
+
+  override def updateOutput(input: Tensor[Float]): Tensor[Float] = {
+    require(input.size(2) == 3)
+    output.resize(input.size(1), 2).zero().addmm(input, w)
+  }
+
+  override def updateGradInput(input: Tensor[Float], gradOutput: Tensor[Float]): Tensor[Float] = {
+    gradInput.resizeAs(input)
+    gradInput.zero().addmm(gradOutput, w.t())
+  }
+}
+```
+
+However, the parameter w here is not trainable. We need override several methods. Here's the code
+example
+```scala
+  private val g = Tensor[Float](3, 2)
+
+  override def accGradParameters(input: Tensor[Float], gradOutput: Tensor[Float]): Unit = {
+    g.addmm(gradOutput, input.t())
+  }
+
+  override def parameters(): (Array[Tensor[Float]], Array[Tensor[Float]]) = {
+    (Array(w), Array(g))
+  }
+
+  override def zeroGradParameters(): Unit = {
+    g.zero()
+  }
+
+  override def reset(): Unit = {
+      w.rand()
+      g.zero()
+  }
+```
+First we introduce a new tensor g, which has same size with parameter w. We call g gradParameter, as
+it stores the gradients which will be used to update w.
```

Comment:
How is w updated using g? Maybe we should add some elaboration here (reference to optimizer related docs).

Reply:
done